### PR TITLE
Porting aux BMS to new version of firmware

### PR DESCRIPTION
**Pull request by badeniran**
Fixing #136

### CODE REVIEW

**Comment by bill-luu** [comment]
Can you split this into 2 commits? One for the auto-generated code, and one for the changes to our actual code? The meaningful changes are buried in the diff now lol

Also, specify the exact version of the firmware in your commit message

File: EpsilonAuxBMS/Core/Src/ReportAuxStatusToCanTask.c [hunk 1]
```diff
@@ -0,0 +1,60 @@
+#include "ReportAuxStatusToCanTask.h"
+
+// Refer to communication protocol
+static const uint32_t AUX_STATUS_FREQ  = 100; // 10Hz
+static const uint32_t AUX_STATUS_STDID  = 0x721U;
+
+void reportAuxStatusToCanTask(void const* arg)
+{
+    // For concurrency with sendHeartbeatTask()
+    osMutexId_t* canHandleMutex = (osMutexId_t*) arg;
```
Comment (badeniran) [inline EpsilonAuxBMS/Core/Src/ReportAuxStatusToCanTask.c]: Diff:
`osMutexId* canHandleMutex = (osMutexId*) arg;`

File: EpsilonAuxBMS/Core/Src/ReportAuxStatusToCanTask.c [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    // One time osDelayUntil initialization
+    uint32_t prevWakeTime = osKernelSysTick();
+
+    for (;;)
+    {
+        osDelayUntil(prevWakeTime + AUX_STATUS_FREQ);
```
Comment (badeniran) [inline EpsilonAuxBMS/Core/Src/ReportAuxStatusToCanTask.c]: Diff:
`osDelayUntil(prevWakeTime + AUX_STATUS_FREQ);`

File: EpsilonAuxBMS/Core/Src/ReportAuxStatusToCanTask.c [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        }
+
+        if (HAL_CAN_GetTxMailboxesFreeLevel(&hcan1) > 0) // Any available Tx mailboxes?
+        {
+            uint8_t data[2];
+            uint32_t mailbox; // Might find a use for mailbox parameter later
+            // (e.g. to implement TxMailbox?CompleteCallback to check a transmission actually completed)
+            // Set CAN Tx header info
+            canTxHeader.StdId = AUX_STATUS_STDID;
+            canTxHeader.DLC = 2;
```
Comment (badeniran) [inline EpsilonAuxBMS/Core/Src/ReportAuxStatusToCanTask.c]: Diff from 28 to this line:
`if (osMutexWait(canHandleMutex, 0) != osOK);`

File: EpsilonAuxBMS/Core/Src/ReportAuxStatusToCanTask.c [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        auxStatus.chargeContactorState = !HAL_GPIO_ReadPin(CHARGE_SENSE_GPIO_Port, CHARGE_SENSE_Pin);
+        auxStatus.dischargeContactorState = !HAL_GPIO_ReadPin(DISCHARGE_SENSE_GPIO_Port, DISCHARGE_SENSE_Pin);
+        auxStatus.highVoltageEnableState = HAL_GPIO_ReadPin(HV_ENABLE_GPIO_Port, HV_ENABLE_Pin);
+
+        if (osMutexAcquire(canHandleMutex, 0) != osOK)
+        {
+            continue;
+        }
+
+        if (HAL_CAN_GetTxMailboxesFreeLevel(&hcan1) > 0) // Any available Tx mailboxes?
```
Comment (badeniran) [inline EpsilonAuxBMS/Core/Src/ReportAuxStatusToCanTask.c]: Diff from line 28 to this line:
`if (osMutexWait(canHandleMutex, 0) != osOK)`

File: EpsilonAuxBMS/Core/Src/ReportAuxStatusToCanTask.c [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        {
+            uint8_t data[2];
+            uint32_t mailbox; // Might find a use for mailbox parameter later
+            // (e.g. to implement TxMailbox?CompleteCallback to check a transmission actually completed)
+            // Set CAN Tx header info
+            canTxHeader.StdId = AUX_STATUS_STDID;
```
Comment (badeniran) [inline EpsilonAuxBMS/Core/Src/ReportAuxStatusToCanTask.c]: Diff:
hcan1.pTxMsg->StdId = AUX_STATUS_STDID;

File: EpsilonAuxBMS/Core/Src/ReportAuxStatusToCanTask.c (hunk 3 above)
Comment (badeniran) [inline EpsilonAuxBMS/Core/Src/ReportAuxStatusToCanTask.c]: Diff:
`hcan1.pTxMsg->DLC = 2;`

File: EpsilonAuxBMS/Core/Src/ReportAuxStatusToCanTask.c [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+            canTxHeader.DLC = 2;
+
+            // Set Data
+            data[0] =
```
Comment (badeniran) [inline EpsilonAuxBMS/Core/Src/ReportAuxStatusToCanTask.c]: Diff:
`hcan1.pTxMsg->Data[0] =`

File: EpsilonAuxBMS/Core/Src/ReportAuxStatusToCanTask.c [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+                auxStatus.commonContactorState |
+                auxStatus.chargeContactorState << 1 |
+                auxStatus.dischargeContactorState << 2 |
+                auxStatus.auxVoltage << 3;
+
+            data[1] =
```
Comment (badeniran) [inline EpsilonAuxBMS/Core/Src/ReportAuxStatusToCanTask.c]: Diff:
`hcan1.pTxMsg->Data[1] =`

File: EpsilonAuxBMS/Core/Src/ReportAuxStatusToCanTask.c [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+                auxStatus.strobeBmsLight |
+                auxStatus.allowCharge << 1 |
+                auxStatus.contactorError << 2 |
+                auxStatus.highVoltageEnableState << 3;
+
+            // Send CAN message
+            if (HAL_CAN_AddTxMessage(&hcan1, &canTxHeader, data, &mailbox) == HAL_OK)
```
Comment (badeniran) [inline EpsilonAuxBMS/Core/Src/ReportAuxStatusToCanTask.c]: Diff:
`if (HAL_CAN_Transmit_IT(&hcan1) == HAL_OK)`

File: EpsilonAuxBMS/Core/Src/ReportHeartbeatToCanTask.c [hunk 9]
```diff
@@ -0,0 +1,43 @@
+#include "ReportHeartbeatToCanTask.h"
+
+// Refer to communications protocol
+static const uint32_t AUX_HEARBEAT_FREQ = 1000; // 1Hz
+static const uint32_t AUX_HEARTBEAT_STDID = 0x720U;
+
+void reportHeartbeatToCanTask(void const* arg)
+{
+    // For concurrency with reportAuxToCanTask()
+    osMutexId_t* canHandleMutex = (osMutexId_t*) arg;
```
Comment (badeniran) [inline EpsilonAuxBMS/Core/Src/ReportHeartbeatToCanTask.c]: Diff:
`osMutexId_t* canHandleMutex = (osMutexId_t*) arg;`

File: EpsilonAuxBMS/Core/Src/ReportHeartbeatToCanTask.c [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    // One time osDelayUntil initialization
+    uint32_t prevWakeTime = osKernelSysTick();
+
+    for (;;)
+    {
+        osDelayUntil(prevWakeTime + AUX_HEARBEAT_FREQ);
```
Comment (badeniran) [inline EpsilonAuxBMS/Core/Src/ReportHeartbeatToCanTask.c]: Diff:
`osDelayUntil(&prevWakeTime, AUX_HEARBEAT_FREQ);`

File: EpsilonAuxBMS/Core/Src/ReportHeartbeatToCanTask.c [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+        if (osMutexAcquire(canHandleMutex, 0) != osOK)
```
Comment (badeniran) [inline EpsilonAuxBMS/Core/Src/ReportHeartbeatToCanTask.c]: Diff:
` if (osMutexWait(canHandleMutex, 0) != osOK)`

File: EpsilonAuxBMS/Core/Src/ReportHeartbeatToCanTask.c [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+        {
+            continue;
+        }
+
+        if (HAL_CAN_GetTxMailboxesFreeLevel(&hcan1) > 0) // Any available Tx mailboxes?
```
Comment (badeniran) [inline EpsilonAuxBMS/Core/Src/ReportHeartbeatToCanTask.c]: Diff:
This if statement was not needed in the previous version

File: EpsilonAuxBMS/Core/Src/ReportAuxStatusToCanTask.c [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    for (;;)
+    {
+        osDelayUntil(prevWakeTime + AUX_STATUS_FREQ);
+
+        auxStatus.commonContactorState = !HAL_GPIO_ReadPin(COMMON_SENSE_GPIO_Port, COMMON_SENSE_Pin);
+        auxStatus.chargeContactorState = !HAL_GPIO_ReadPin(CHARGE_SENSE_GPIO_Port, CHARGE_SENSE_Pin);
+        auxStatus.dischargeContactorState = !HAL_GPIO_ReadPin(DISCHARGE_SENSE_GPIO_Port, DISCHARGE_SENSE_Pin);
+        auxStatus.highVoltageEnableState = HAL_GPIO_ReadPin(HV_ENABLE_GPIO_Port, HV_ENABLE_Pin);
+
+        if (osMutexAcquire(canHandleMutex, 0) != osOK)
```
Comment (badeniran) [inline EpsilonAuxBMS/Core/Src/ReportAuxStatusToCanTask.c]: Diff:
` if (osMutexWait(canHandleMutex, 0) != osOK)`

File: EpsilonAuxBMS/Core/Src/ReportHeartbeatToCanTask.c [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+        {
+            uint8_t data[1];
+            uint32_t mailbox;
+            // Set CAN message address
+            canTxHeader.StdId = AUX_HEARTBEAT_STDID;
```
Comment (badeniran) [inline EpsilonAuxBMS/Core/Src/ReportHeartbeatToCanTask.c]: Diff:
 `hcan1.pTxMsg->StdId = AUX_HEARTBEAT_STDID;`

File: EpsilonAuxBMS/Core/Src/ReportHeartbeatToCanTask.c [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+            canTxHeader.DLC = 1;
```
Comment (badeniran) [inline EpsilonAuxBMS/Core/Src/ReportHeartbeatToCanTask.c]: Diff:
`hcan1.pTxMsg->DLC = 1;`

File: EpsilonAuxBMS/Core/Src/ReportHeartbeatToCanTask.c [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+            // Always 1
+            data[0] = 1;
```
Comment (badeniran) [inline EpsilonAuxBMS/Core/Src/ReportHeartbeatToCanTask.c]: Diff:
`hcan1.pTxMsg->Data[0] = 1;`

File: EpsilonAuxBMS/Core/Src/ReportHeartbeatToCanTask.c [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+
+            // Send CAN message
+            if (HAL_CAN_AddTxMessage(&hcan1, &canTxHeader, data, &mailbox) == HAL_OK)
```
Comment (badeniran) [inline EpsilonAuxBMS/Core/Src/ReportHeartbeatToCanTask.c]: Diff:
`if (HAL_CAN_Transmit_IT(&hcan1) == HAL_OK)`

**Pull request closed without merging**